code/game/objects/effects/portals.dm

```diff
@@ -1,35 +1,102 @@
 /obj/effect/portal
 	name = "portal"
-	desc = "Looks unstable. Best to test it with the clown."
+	desc = "There's a hole in the sky... through which I can fly."
 	icon = 'icons/obj/stationobjs.dmi'
-	icon_state = "portal"
-	density = 1
+	icon_state = "portal0"
+	density = 0
 	unacidable = 1//Can't destroy energy portals.
-	var/failchance = 5
-	var/obj/item/target = null
-	var/obj/item/weapon/hand_tele/creator = null
+	var/obj/target = null
+	var/obj/item/weapon/creator = null
 	anchored = 1.0
 	w_type=NOT_RECYCLABLE
+	var/undergoing_deletion = 0
 
+	var/list/exit_beams = list()
+
+/obj/effect/portal/attack_hand(var/mob/user)
+	spawn()
+		src.teleport(user)
+
+/obj/effect/portal/attackby(obj/item/weapon/O as obj, mob/user as mob)
+	if(O == creator)
+		user << "<span class='warning'>You close the portal prematurely.</span>"
+		qdel(src)
+	else
+		spawn()
+			src.teleport(user)
+/*
 /obj/effect/portal/Bumped(mob/M as mob|obj)
-	spawn(0)
+	spawn()
 		src.teleport(M)
-
+*/
 /obj/effect/portal/Crossed(AM as mob|obj)
-	spawn(0)
+	if(istype(AM,/obj/item/projectile/beam))
+		var/obj/item/projectile/beam/B = AM
+		B.wait = 1
+	spawn()
 		src.teleport(AM)
 
-/obj/effect/portal/New()
+/obj/effect/portal/CanPass(atom/movable/mover, turf/target, height=1.5, air_group = 0)
+	if(istype(mover,/obj/effect/beam))
+		return 0
+	else
+		return ..()
+
+/obj/effect/portal/New(turf/loc,var/lifespan=300)
 	..()
-	spawn(300)
+	playsound(loc,'sound/effects/portal_open.ogg',60,1)
+	spawn(lifespan)
 		qdel(src)
 
 /obj/effect/portal/Destroy()
-	..()
-	target = null
+	if(undergoing_deletion)
+		return
+	undergoing_deletion = 1
+	playsound(loc,'sound/effects/portal_close.ogg',60,1)
+
+	purge_beams()
+
+	if(target)
+		if(istype(target,/obj/effect/portal) && !istype(creator,/obj/item/weapon/gun/portalgun))
+			qdel(target)
+		target = null
 	if(creator)
-		creator.portals -= src
-		creator = null
+		if(istype(creator,/obj/item/weapon/hand_tele))
+			var/obj/item/weapon/hand_tele/H = creator
+			H.portals -= src
+			creator = null
+		else if(istype(creator,/obj/item/weapon/gun/portalgun))
+			var/obj/item/weapon/gun/portalgun/P = creator
+			if(src == P.blue_portal)
+				P.blue_portal = null
+				P.sync_portals()
+			else if(src == P.red_portal)
+				P.red_portal = null
+				P.sync_portals()
+
+	var/datum/effect/effect/system/spark_spread/aeffect = new
+	aeffect.set_up(5, 1, loc)
+	aeffect.start()
+	..()
+
+/obj/effect/portal/cultify()
+	return
+
+/obj/effect/portal/singuloCanEat()
+	return 0
+
+/obj/effect/portal/singularity_act()
+	return
+
+/obj/effect/portal/singularity_pull()
+	return
+
+/obj/effect/portal/proc/blend_icon(var/obj/effect/portal/P)

```

Comment:
@DeityLink consider caching blended states after they're blended for the first time though.